src/mbgl/shader/shader.cpp

```diff
@@ -26,26 +26,53 @@ Shader::Shader(const char *name_, const GLchar *vertSource, const GLchar *fragSo
     if (!binaryFileName.empty() && (gl::ProgramBinary != nullptr)) {
-        std::ifstream binaryFile(binaryFileName, std::ios::in | std::ios::binary);
+        try {
+            std::ifstream binaryFile(binaryFileName, std::ios::in | std::ios::binary);
+            binaryFile.exceptions(std::ifstream::failbit | std::ifstream::badbit | std::ifstream::eofbit);
 
-        GLsizei binaryLength;
-        GLenum binaryFormat;
-        binaryFile.read(reinterpret_cast<char *>(&binaryLength), sizeof(binaryLength));
-        binaryFile.read(reinterpret_cast<char *>(&binaryFormat), sizeof(binaryFormat));
+            GLsizei binaryLength;
+            GLenum binaryFormat;
+            binaryFile.read(reinterpret_cast<char *>(&binaryLength), sizeof(binaryLength));
+            binaryFile.read(reinterpret_cast<char *>(&binaryFormat), sizeof(binaryFormat));
 
-        std::unique_ptr<char[]> binary = mbgl::util::make_unique<char[]>(binaryLength);
-        binaryFile.read(binary.get(), binaryLength);
+            GLint numBinaryFormats;
+            MBGL_CHECK_ERROR(glGetIntegerv(GL_NUM_PROGRAM_BINARY_FORMATS, &numBinaryFormats));
 
-        MBGL_CHECK_ERROR(gl::ProgramBinary(program, binaryFormat, binary.get(), binaryLength));
+            std::unique_ptr<GLenum[]> validBinaryFormats = mbgl::util::make_unique<GLenum[]>(numBinaryFormats);
+            MBGL_CHECK_ERROR(glGetIntegerv(GL_PROGRAM_BINARY_FORMATS, reinterpret_cast<GLint *>(validBinaryFormats.get())));
 
-        // Check if the binary was valid
-        GLint status;
-        MBGL_CHECK_ERROR(glGetProgramiv(program, GL_LINK_STATUS, &status));
-        if (status == GL_TRUE) {
-            skipCompile = true;
-        }
+            bool validBinaryFormat = false;
+            for (GLint i = 0; i < numBinaryFormats; i++) {
+                if (validBinaryFormats[i] == binaryFormat) {
+                    validBinaryFormat = true;
+                }
+            }
+            if (!validBinaryFormat) {
+                throw std::runtime_error("Trying load program binary with an invalid binaryFormat!");
+            }
+
+            if (binaryLength == 0) {
+                throw std::runtime_error("Trying load program binary with a zero length binary!");
+            }
+
+            std::unique_ptr<char[]> binary = mbgl::util::make_unique<char[]>(binaryLength);
+            binaryFile.read(binary.get(), binaryLength);
 
-        binaryFile.close();
+            MBGL_CHECK_ERROR(gl::ProgramBinary(program, binaryFormat, binary.get(), binaryLength));
+
+            // Check if the binary was valid
+            GLint status;
+            MBGL_CHECK_ERROR(glGetProgramiv(program, GL_LINK_STATUS, &status));
+            if (status == GL_TRUE) {
+                skipCompile = true;
+            }
+
+        } catch(std::exception e) {
+            Log::Error(Event::Shader, "Loading binary shader failed!");
+
+            // Delete the bad file
+            std::remove(binaryFileName.c_str());
```

Comment:
Which [`std::remove`](http://en.cppreference.com/w/cpp/algorithm/remove) is this? I'm surprised a one-argument call compiles.